### ISSUE TITLE
Strange bug

### ISSUE THREAD

**Issue by noughtmare** [xnning/EvEff#1]
I've been trying this library out to write a parser effect by combining state and choice. I've run into a very strange bug. Here is my code:

```Haskell
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE TypeOperators, FlexibleContexts, Rank2Types #-}

import           Control.Ev.Eff
import           Control.Ev.Util
import           Control.Applicative

satisfy :: (State [s] :? e, Choose :? e) => (s -> Bool) -> Eff e s
satisfy p = perform get () >>= \case
  (x : xs) | p x -> x <$ perform put xs
  _              -> empty

main :: IO ()
main = print . runEff . chooseAll . state "abc" $ satisfy (== 'a') *> satisfy (== 'c') <|> satisfy (== 'c')
```

This prints `"\140412572307544"`.

**Comment by noughtmare** [comment]
~~I just tested this again with the master branch on GitHub here and this problem no longer occurs. Can you please upload the fixed version to Hackage?~~

EDIT: Actually, it only works correctly if I load this library in the repl. It doesn't work correctly when I import this library in another package.

**Comment by xnning** [comment]
Thanks for the bug reporting!

Sorry I just noticed this issue. I will take a look (together with #2 ) and let you know.

**Comment by xnning** [comment]
Haven't figured out why but there seems to be something wrong with `<|>`.

A minimal example that reproduces the weird behavior is

```
get1 :: Eff (State Int :* e) Int
get1 = perform get ()

test = runEff $ chooseAll $ state (1::Int) $ (get1 <|> get1)
```

**Comment by xnning** [comment]
Fixed by eeee566.

The reason is that with `-O2`, GHC aggressively optimizes the implementation, which also inlines the unsafeIO part for markers. The result is that comparing handlers always returns true, so in the previous example, the `state` handler is used to handle `choose` (raised by `<|>`) and things go wrong.

Adding `NOINLINE` fixes the issue.
